src/proto/streams/recv.rs

```diff
@@ -262,10 +264,21 @@ impl<B, P> Recv<B, P>
 
         // Ensure that there is enough capacity on the connection before acting
         // on the stream.
-        if self.flow.window_size() < sz || stream.recv_flow.window_size() < sz {
+        if self.flow.window_size() < sz {
             return Err(RecvError::Connection(FlowControlError));
         }
 
+        // The spec gives us the option of returning either a connection error
+        // or stream error when a stream overflows its window (but not the
+        // connection). So, we opt to send a stream error.
+        if stream.recv_flow.window_size() < sz {
```

Comment:
If you use a stream level error here, then `sz` has to be applied against the connection level flow control.

> A receiver that receives a flow-controlled frame MUST always account for its contribution against the connection flow-control window, unless the receiver treats this as a connection error (Section 5.4.1). This is necessary even if the frame is in error. The sender counts the frame toward the flow-control window, but if the receiver does not, the flow-control window at the sender and receiver can become different.

This relates to #32 and generally speaking there probably are existing bugs related to this in the repo.